### ISSUE TITLE
swe_pheno output

### ISSUE THREAD

**Issue by vreijs** [rstub/swephR#20]
Here are example using swetest.exe (form SE author Alois):
```
Examples:
Swiss Ephemeris
swetest -p1e -fPL+-*/= -b1.1.2018 -eswe
date (dmy) 1.1.2018 greg.   0:00:00 TT        version 2.07.02
TT:  2458119.500000000
Epsilon (true)    23°26' 5.6136
Nutation          -0° 0'11.5628   -0° 0' 7.3618
name            long.          phase phase elong diamet magn
Moon              84°47'14.6888   16°10'23.8764      0.980211798  163°47'15.7090    0°33'26.2108   -12.46m
Moshier:
swetest -p1e -fPL+-*/= -b1.1.2018 -emos
date (dmy) 1.1.2018 greg.   0:00:00 TT        version 2.07.02
TT:  2458119.500000000
Epsilon (true)    23°26' 5.6136
Nutation          -0° 0'11.5628   -0° 0' 7.3618
name            long.          phase phase elong diamet magn
Moon              84°47'14.9997   16°10'23.4491      0.980212087  163°47'16.1424    0°33'26.2036   -12.46m

JPL:
swetest -p1e -fPL+-*/= -b1.1.2018 -ejpl
date (dmy) 1.1.2018 greg.   0:00:00 TT        version 2.07.02
TT:  2458119.500000000
Epsilon (true)    23°26' 5.6136
Nutation          -0° 0'11.5628   -0° 0' 7.3618
name            long.          phase phase elong diamet magn
Moon              84°47'14.6891   16°10'23.8665      0.980211805  163°47'15.7190    0°33'26.2108   -12.46m

Because Moshier Moon can differ slightly (0.3 arcsec in this example from JPL/Swisseph, there are of course small differences in phase and elongation.
```

I will try to see if I can reproduce these and if Moshier changes when the SWIEPH is loaded...

**Comment by vreijs** [comment]
At this moment I am not able to repeat what I (we?) first saw: that Moshier results changed when changing the set_ephe_path to SWIEPH data. See below tests:

```
> swe_set_ephe_path('') 
> swe_pheno_ut(2458119.5,1,4) 
$`return` [1] 0  
$attr  [1]  16.162033224870228   0.980239173478929 163.798990566277666   0.557281061427825 -12.457146749869864   1.022548960400099  [7]   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000 [13]   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000 [19]   0.000000000000000   0.000000000000000  
$serr [1] ""  
> swe_pheno_ut(2458119.5,1,2) 
$`return` [1] 0  
$attr  [1]  16.162033224870228   0.980239173478929 163.798990566277666   0.557281061427825 -12.457146749869864   1.022548960400099  [7]   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000 [13]   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000 [19]   0.000000000000000   0.000000000000000  
$serr [1] "SwissEph file 'sepl_18.se1' not found in PATH '\\sweph\\ephe\\' \nusing Moshier eph.; "  
> swe_set_ephe_path('C:\\Users\\VRHome\\Documents\\R\\swephRdata\\inst\\ephemeris') 
> swe_pheno_ut(2458119.5,1,4) 
$`return` [1] 0  
$attr  [1]  16.162033224870228   0.980239173478929 163.798990566277666   0.557281061427825 -12.457146749869864   1.022548960400099  [7]   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000 [13]   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000 [19]   0.000000000000000   0.000000000000000  
$serr [1] ""  
> swe_pheno_ut(2458119.5,1,2) 
$`return` [1] 0  
$attr  [1]  16.162151844106774   0.980238885339545 163.798870275611733   0.557283055225551 -12.457151426610626   1.022552592970599  [7]   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000 [13]   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000 [19]   0.000000000000000   0.000000000000000  $serr [1] ""
--
 
> | >
>


```

**Comment by vreijs** [comment]
By the way, I get now in any case the same results as SE author (when using the correct time).

**Comment by rstub** [comment]
I can still reproduce the issue for the test-day in the past:

``` r
library(swephR)
swe_set_ephe_path("does/not/exist")
swe_pheno_ut(1234567,1,4)
#> $return
#> [1] 0
#> 
#> $attr
#>  [1] 149.63072160   0.06860756  30.29443453   0.51804332  -6.82356145
#>  [6]   0.95056874   0.00000000   0.00000000   0.00000000   0.00000000
#> [11]   0.00000000   0.00000000   0.00000000   0.00000000   0.00000000
#> [16]   0.00000000   0.00000000   0.00000000   0.00000000   0.00000000
#> 
#> $serr
#> [1] ""
swe_pheno_ut(1234567,1,2)
#> $return
#> [1] 0
#> 
#> $attr
#>  [1] 149.63072160   0.06860756  30.29443453   0.51804332  -6.82356145
#>  [6]   0.95056874   0.00000000   0.00000000   0.00000000   0.00000000
#> [11]   0.00000000   0.00000000   0.00000000   0.00000000   0.00000000
#> [16]   0.00000000   0.00000000   0.00000000   0.00000000   0.00000000
#> 
#> $serr
#> [1] "SwissEph file 'seplm18.se1' not found in PATH 'does/not/exist/' \nusing Moshier eph.; "
swe_set_ephe_path(system.file("ephemeris", package = "swephRdata"))
swe_pheno_ut(1234567,1,4)
#> $return
#> [1] 0
#> 
#> $attr
#>  [1] 149.60020709   0.06874225  30.32487839   0.51802614  -6.82591793
#>  [6]   0.95053725   0.00000000   0.00000000   0.00000000   0.00000000
#> [11]   0.00000000   0.00000000   0.00000000   0.00000000   0.00000000
#> [16]   0.00000000   0.00000000   0.00000000   0.00000000   0.00000000
#> 
#> $serr
#> [1] ""
swe_pheno_ut(1234567,1,2)
#> $return
#> [1] 0
#> 
#> $attr
#>  [1] 149.62582456   0.06862917  30.29932000   0.51803853  -6.82393114
#>  [6]   0.95055991   0.00000000   0.00000000   0.00000000   0.00000000
#> [11]   0.00000000   0.00000000   0.00000000   0.00000000   0.00000000
#> [16]   0.00000000   0.00000000   0.00000000   0.00000000   0.00000000
#> 
#> $serr
#> [1] ""
```

There is indeed no problem for a current date:

``` r
swe_set_ephe_path("does/not/exist")
swe_pheno_ut(2458119.5,1,4)
#> $return
#> [1] 0
#> 
#> $attr
#>  [1]  16.1620332   0.9802392 163.7989906   0.5572811 -12.4571467
#>  [6]   1.0225490   0.0000000   0.0000000   0.0000000   0.0000000
#> [11]   0.0000000   0.0000000   0.0000000   0.0000000   0.0000000
#> [16]   0.0000000   0.0000000   0.0000000   0.0000000   0.0000000
#> 
#> $serr
#> [1] ""
swe_pheno_ut(2458119.5,1,2)
#> $return
#> [1] 0
#> 
#> $attr
#>  [1]  16.1620332   0.9802392 163.7989906   0.5572811 -12.4571467
#>  [6]   1.0225490   0.0000000   0.0000000   0.0000000   0.0000000
#> [11]   0.0000000   0.0000000   0.0000000   0.0000000   0.0000000
#> [16]   0.0000000   0.0000000   0.0000000   0.0000000   0.0000000
#> 
#> $serr
#> [1] "SwissEph file 'sepl_18.se1' not found in PATH 'does/not/exist/' \nusing Moshier eph.; "
swe_set_ephe_path(system.file("ephemeris", package = "swephRdata"))
swe_pheno_ut(2458119.5,1,4)
#> $return
#> [1] 0
#> 
#> $attr
#>  [1]  16.1620332   0.9802392 163.7989906   0.5572811 -12.4571467
#>  [6]   1.0225490   0.0000000   0.0000000   0.0000000   0.0000000
#> [11]   0.0000000   0.0000000   0.0000000   0.0000000   0.0000000
#> [16]   0.0000000   0.0000000   0.0000000   0.0000000   0.0000000
#> 
#> $serr
#> [1] ""
swe_pheno_ut(2458119.5,1,2)
#> $return
#> [1] 0
#> 
#> $attr
#>  [1]  16.1621518   0.9802389 163.7988703   0.5572831 -12.4571514
#>  [6]   1.0225526   0.0000000   0.0000000   0.0000000   0.0000000
#> [11]   0.0000000   0.0000000   0.0000000   0.0000000   0.0000000
#> [16]   0.0000000   0.0000000   0.0000000   0.0000000   0.0000000
#> 
#> $serr
#> [1] ""
```

Created on 2018-09-10 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).


I will tr to reproduce this using `swetest`.

**Comment by vreijs** [comment]
mmmhhh, I think it is the DeltaT that does it (the earlier example from SE author, is too close to deltat=0 perhaps;-).
So if I take our original JD number (1234567). I get:

```
> swe_set_ephe_path('C:\\Users\\VRHome\\Documents\\R\\swephRdata\\inst\\ephemeris')
> swe_pheno_ut(1234567,1,4)
$`return`
[1] 0

$attr
 [1] 149.600207099878219   0.068742250850754  30.324878387827614   0.518026140482446  -6.825917933442083   0.950537249978976
 [7]   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000
[13]   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000   0.000000000000000
[19]   0.000000000000000   0.000000000000000

$serr
[1] ""

> swe_deltat_ex(1234567, 4)
$`deltat`
[1] 0.366039979375346

$serr
[1] ""

> swe_deltat(1234567)
[1] 0.368543449981732
> swe_set_ephe_path('')
> swe_pheno_ut(1234567,1,4)
$`return`
[1] 0

$attr
 [1] 149.6307215971654045   0.0686075615100671  30.2944345224698139   0.5180433249746146  -6.8235614464226648   0.9505687417907285
 [7]   0.0000000000000000   0.0000000000000000   0.0000000000000000   0.0000000000000000   0.0000000000000000   0.0000000000000000
[13]   0.0000000000000000   0.0000000000000000   0.0000000000000000   0.0000000000000000   0.0000000000000000   0.0000000000000000
[19]   0.0000000000000000   0.0000000000000000

$serr
[1] ""

> swe_deltat(1234567)
[1] 0.366039979375346
> 

> swe_set_ephe_path('C:\\Users\\VRHome\\Documents\\R\\swephRdata\\inst\\ephemeris') > swe_pheno_ut(1234566.5,1,4) $`return` [1] 0  $attr  [1] 155.7382315846214169   0.0441611681600523  24.2013703634234965  [4]   0.5214504178825498  -6.3310302088048935   0.9568120803463117  [7]   0.0000000000000000   0.0000000000000000   0.0000000000000000 [10]   0.0000000000000000   0.0000000000000000   0.0000000000000000 [13]   0.0000000000000000   0.0000000000000000   0.0000000000000000 [16]   0.0000000000000000   0.0000000000000000   0.0000000000000000 [19]   0.0000000000000000   0.0000000000000000  $serr [1] ""  > swe_deltat(1234566.5) [1] 0.368543774639431 > swe_set_ephe_path('') > swe_pheno_ut(1234566.5,1,4) $`return` [1] 0  $attr  [1] 155.7691800811143423   0.0440502585450099  24.1704964555454183  [4]   0.5214674651329871  -6.3284255320279676   0.9568433167170607  [7]   0.0000000000000000   0.0000000000000000   0.0000000000000000 [10]   0.0000000000000000   0.0000000000000000   0.0000000000000000 [13]   0.0000000000000000   0.0000000000000000   0.0000000000000000 [16]   0.0000000000000000   0.0000000000000000   0.0000000000000000 [19]   0.0000000000000000   0.0000000000000000  $serr [1] ""  > swe_deltat(1234566.5) [1] 0.366040301947761
--
 
> | >
>


```
If I repeat this in swetest I get the same swetest results with or without SWIEPH. But swetest does always use swe_pheno. And I think swetest does a swe_deltat_ex to get the deltat, while I think that swe_pheno_ut does a swe_deltat... Will ask SE author.

**Comment by vreijs** [comment]
I think I found the issue (I think our messages crossed). In the meantime,
I have asked the SE author if it is indeed the deltat value used in
swe_pheno_ut.
swetest is correct.

**Comment by rstub** [comment]
I have managed to reduce this to a simple C program. If there is any doubt upstream, you can forward it:

```c
#include <stdio.h>
#include "swephexp.h"

int main(int argc, char *argv[]) {
  if (argc < 2)
    swe_set_ephe_path(NULL);
  else
    swe_set_ephe_path(argv[1]);

  char version[AS_MAXCH];
  swe_version(&version[0]);
  printf("Version: %s\n", version);

  double tjd = 1234567;
  int32 ipl = 1;
  int32 iflag = 4;
  double attr[20];
  char serr[AS_MAXCH];

  int32 rc = swe_pheno_ut(tjd, ipl, iflag, &attr[0], &serr[0]);
  printf("RC: %i; message: %s\n", rc, serr); 
  printf("Result UT: %f, %f, %f, %f, %f\n", attr[0], attr[1], attr[2], attr[3], attr[4], attr[5]);  

  rc = swe_pheno(tjd, ipl, iflag, &attr[0], &serr[0]);
  printf("RC: %i; message: %s\n", rc, serr); 
  printf("Result ET: %f, %f, %f, %f, %f\n", attr[0], attr[1], attr[2], attr[3], attr[4], attr[5]);  
  return 0;
}
```

Result:
```

ralf@lewis:~/devel/swephR/external/src$ ./pheno 
Version: 2.07.01
RC: 0; message: 
Result UT: 149.630722, 0.068608, 30.294435, 0.518043, -6.823561
RC: 0; message: 
Result ET: 154.116100, 0.050160, 25.819610, 0.520553, -6.465983
ralf@lewis:~/devel/swephR/external/src$ ./pheno ~/devel/se-data/ephe/
Version: 2.07.01
RC: 0; message: 
Result UT: 149.600207, 0.068742, 30.324878, 0.518026, -6.825918
RC: 0; message: 
Result ET: 154.116100, 0.050160, 25.819610, 0.520553, -6.465983
```

The crucial thing is the the ET results stay the same, even when the SE data is available. The UT data changes, though.

**Comment by rstub** [comment]
A minimal patch that fixes the issue in my limited testing:

```
diff --git a/external/src/swecl.c b/external/src/swecl.c
index 6b7dfde..17133d6 100644
--- a/external/src/swecl.c
+++ b/external/src/swecl.c
@@ -3992,7 +3992,7 @@ int32 CALL_CONV swe_pheno(double tjd, int32 ipl, int32 iflag, double *attr, char
   }
   if (*serr2 != '\0' && serr != NULL)
     strcpy(serr, serr2);
-  return OK;
+  return iflag;
 }
 
 int32 CALL_CONV swe_pheno_ut(double tjd_ut, int32 ipl, int32 iflag, double *attr, char *serr)
```

Instead of always returning `OK`, which is 0, `swe_pheno` now follows the usual pattern to return the flags that where actually used.

**Comment by vreijs** [comment]
I had deducted the same, which I put on the swissephe e-group this morning.
No confirmation yet received. I though did not provided the solution,
thanks for this. I will await their feedback for another day, otherwise I
will give the give of yours.

**Comment by vreijs** [comment]
I had provided similar results to them already. Again will wait until
tomorrow, I hope they also resolved it in the 'morther' source (and I hope
a similar thing is not happening in other functions).

**Comment by rstub** [comment]
Ok. For the time being I have committed my patch in #21. This also helped to bring the CI to "green".

**Comment by vreijs** [comment]
Thanks for changing the C-code of swe_pheno! I just tested it and it looks
great. I did not yet got the reaction on the copyright question, but I
think in the meantime we should use what you propose: "Authors of the Swiss
Ephemeris" and install their LICENSE.TXT along with the package (both the
R-code and the SWIEPH data).
In that way we obey SE's own rules, IMHO. If they did not react in before
the weekend, I will mention on the e-group that that is what we do.
